### ISSUE TITLE
strange relative path listed for saved files

### ISSUE THREAD

**Issue by mattcieslak** [datalad/datalad#5175]
Hi! I'm running into some strange behavior when running datalad on a compute cluster. Here is what an example shell session looks like (running from my home directory)

```
$ mkdir mytest
$ cd mytest
$ datalad create --description test .
[INFO   ] Creating a new annex repo at /cbica/projects/RBC/mytest 
[INFO   ] Scanning for unlocked files (this may take some time) 
$ echo content > testfile
$ datalad save -m 'initial commit' 
add(ok):../../../../gpfs/fs001/cbica/projects/RBC/mytest/ /cbica/projects/RBC/mytest/testfile (file)
save(ok):../../../../gpfs/fs001/cbica/projects/RBC/mytest/ /cbica/projects/RBC/mytest (dataset)
action summary:
  add (ok: 1)
  save (ok: 1)

```

datalad version 0.13.5
I was wondering why I might be seeing `../../../..` in the path. Is this something I should worry about? Thanks in advance

**Comment by kyleam** [comment]
Huh, that does look off, I think.  The path reporting can be a bit strange and misleading (see gh-3442 and a work-in-progress fix at gh-4454), but I don't see why it would be for the command snippets you show.  Can you give a description of the larger directory structure (in particularly nested git directories and directory symlinks)?

**Comment by mattcieslak** [comment]
Thanks for the quick reply!
/gpfs is an nfs drive mounted from `/dev/mapper/vg_local-lv_root`

The dataset is not nested in any other dataset
```
[rbc@cubic-login2 ~]$ git status
fatal: not a git repository (or any parent up to mount point /gpfs)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
[rbc@cubic-login2 another_test]$ mkdir another_test 
[rbc@cubic-login2 another_test]$ cd another_test
[rbc@cubic-login2 another_test]$ datalad create --description "another test" -c text2git ./
[INFO   ] Creating a new annex repo at /cbica/projects/RBC/another_test 
[INFO   ] Scanning for unlocked files (this may take some time) 
[INFO   ] Running procedure cfg_text2git 
[INFO   ] == Command start (output follows) ===== 
[INFO   ] == Command exit (modification check follows) ===== 
create(ok): /cbica/projects/RBC/another_test (dataset)
[rbc@cubic-login2 another_test]$ echo data >> testfile
[rbc@cubic-login2 another_test]$ datalad save -m "adding a text file"
add(ok):../../../../gpfs/fs001/cbica/projects/RBC/another_test/ /cbica/projects/RBC/another_test/testfile (file)
save(ok):../../../../gpfs/fs001/cbica/projects/RBC/another_test/ /cbica/projects/RBC/another_test (dataset)
action summary:
  add (ok: 1)
  save (ok: 1)
[rbc@cubic-login2 another_test]$ datalad wtf
# WTF
## configuration <SENSITIVE, report disabled by configuration>
## datalad 
  - full_version: 0.13.5
  - version: 0.13.5
## dataset 
  - id: fbbb5249-88cc-42c5-b260-74c3c6a152e9
  - metadata: <SENSITIVE, report disabled by configuration>
  - path: /cbica/projects/RBC/another_test
  - repo: AnnexRepo

```

The actual files are 
```
[rbc@cubic-login2 another_test]$ find . | sort
.
./.datalad
./.datalad/config
./.datalad/.gitattributes
./.git
./.git/annex
./.git/annex/gitqueue.lck
./.git/annex/index
./.git/annex/index.lck
./.git/annex/journal
./.git/annex/journal.lck
./.git/annex/keysdb.lck
./.git/annex/othertmp.lck
./.git/annex/sentinal
./.git/annex/sentinal.cache
./.git/annex/smudge.lck
./.gitattributes
./.git/COMMIT_EDITMSG
./.git/config
./.git/description
./.git/HEAD
./.git/hooks
./.git/hooks/applypatch-msg.sample
./.git/hooks/commit-msg.sample
./.git/hooks/fsmonitor-watchman.sample
./.git/hooks/post-checkout
./.git/hooks/post-merge
./.git/hooks/post-receive
./.git/hooks/post-update.sample
./.git/hooks/pre-applypatch.sample
./.git/hooks/pre-commit
./.git/hooks/pre-commit.sample
./.git/hooks/pre-merge-commit.sample
./.git/hooks/prepare-commit-msg.sample
./.git/hooks/pre-push.sample
./.git/hooks/pre-rebase.sample
./.git/hooks/pre-receive.sample
./.git/hooks/update.sample
./.git/index
./.git/info
./.git/info/attributes
./.git/info/exclude
./.git/logs
./.git/logs/HEAD
./.git/logs/refs
./.git/logs/refs/heads
./.git/logs/refs/heads/git-annex
./.git/logs/refs/heads/master
./.git/objects
./.git/objects/12
./.git/objects/12/69488f7fb1f4b56a8c0e5eb48cecbfadfa9219
./.git/objects/16
./.git/objects/16/a4c23200cb0f76b0f3279800c086babca8c5db
./.git/objects/33
./.git/objects/33/8ad54b72957a763b5b52acdbfcb817780e0eee
./.git/objects/4b
./.git/objects/4b/825dc642cb6eb9a060e54bf8d69288fbee4904
./.git/objects/4d
./.git/objects/4d/cb85be8636a2167645488be47289fb8de98067
./.git/objects/51
./.git/objects/51/9a1f93aa8b2965212e9287c6d8943a43952cdb
./.git/objects/56
./.git/objects/56/07e7f7b045b96d20ff7523f6dddb962b203e20
./.git/objects/60
./.git/objects/60/043727424fdb8460d445c1b3e9990ab598ffdd
./.git/objects/6f
./.git/objects/6f/6e51e69f6c97aad2030bb97488a193e67a90f5
./.git/objects/80
./.git/objects/80/8b082abbc690b9c64f5eeb96513fe416070c94
./.git/objects/8a
./.git/objects/8a/ac57d8f70e0fa64aad2d1025cb17f01c0466c1
./.git/objects/97
./.git/objects/97/1df972c7637cb97fea21c4f5221e74d9fe0ccd
./.git/objects/9c
./.git/objects/9c/7d504362403baf2ce42a2d7769462d0c516d5b
./.git/objects/b5
./.git/objects/b5/40820107ca5718dc0f196a08edf3729239bfef
./.git/objects/bc
./.git/objects/bc/9c0407aa1ae84ab8e6ac9c19ce457becbda934
./.git/objects/c3
./.git/objects/c3/aaefef9a2470b31ba9213350046ff7cde75737
./.git/objects/ce
./.git/objects/ce/7d0936961da6253446149441364fecf4098e1f
./.git/objects/d4
./.git/objects/d4/d92240f474cf4205e0966c31a5405d1768b209
./.git/objects/e6
./.git/objects/e6/9de29bb2d1d6434b8b29ae775ad8c2e48c5391
./.git/objects/info
./.git/objects/pack
./.git/refs
./.git/refs/heads
./.git/refs/heads/git-annex
./.git/refs/heads/master
./.git/refs/tags
./testfile
```

Is there any other information that might be helpful? I'm worried that when I clone from this repository it will try to unpack data to directories that don't exist for the cloner.

**Comment by kyleam** [comment]
> mytest/ /cbica

This doesn't necessarily explain the larger issue, but I suspect this space is coming from mishandling in the default result renderer.

https://github.com/datalad/datalad/blob/937a95c1138074acf63ad490e4e129468a3c89da/datalad/interface/utils.py#L500-L503

That spurious space has been fixed along with a few other things in 025bc2ce0 (BF+TST: Robustify default_result_renderer(), 2020-11-03, gh-5121).  If you're able to, it might be worth checking what the those results look like on datalad's maint branch, though I'd guess it'll still be off.

The output of `datalad -f json_pp status` might also give a hint at what's going on.

---

For reference, here's what I'd expect the output to look like on 0.13.5.

<details>
<summary>expected 0.13.5 output</summary>

```sh
set -eu
cd "$(mktemp -d "${TMPDIR:-/tmp}"/dl-XXXXXXX)"
datalad --version
datalad create --description test .
echo content > testfile
datalad save -m 'initial commit' 
```

```
datalad 0.13.5
[INFO   ] Creating a new annex repo at /tmp/dl-gwW3Gai
[INFO   ] Scanning for unlocked files (this may take some time)
create(ok): /tmp/dl-gwW3Gai (dataset)
add(ok): /tmp/dl-gwW3Gai/testfile (file)
save(ok): /tmp/dl-gwW3Gai (dataset)
action summary:
  add (ok: 1)
  save (ok: 1)
```

</details>

**Comment by kyleam** [comment]
> Is there any other information that might be helpful?

Sorry, our posts crossed.  `datalad -f json_pp status` would be good for seeing what the reference dataset is.  You could also redo your `create`/`save` snippet with `-f json_pp`.  (`-f json_pp` changes the result renderer.)

> I'm worried that when I clone from this repository it will try to unpack data to directories that don't exist for the cloner.

I'm hopeful it's just the result renderer (which is purely a cosmetic thing), but, yeah, it'd be good to be sure.

**Comment by mattcieslak** [comment]
The json_pp looks good, I think:

```
[rbc@cubic-login2 another_test]$ datalad -f json_pp status
{
  "action": "status",
  "bytesize": 125,
  "gitshasum": "b540820107ca5718dc0f196a08edf3729239bfef",
  "parentds": "/cbica/projects/RBC/another_test",
  "path": "/cbica/projects/RBC/another_test/.datalad/.gitattributes",
  "prev_gitshasum": "b540820107ca5718dc0f196a08edf3729239bfef",
  "refds": "/cbica/projects/RBC/another_test",
  "state": "clean",
  "status": "ok",
  "type": "file"
}
{
  "action": "status",
  "bytesize": 63,
  "gitshasum": "338ad54b72957a763b5b52acdbfcb817780e0eee",
  "parentds": "/cbica/projects/RBC/another_test",
  "path": "/cbica/projects/RBC/another_test/.datalad/config",
  "prev_gitshasum": "338ad54b72957a763b5b52acdbfcb817780e0eee",
  "refds": "/cbica/projects/RBC/another_test",
  "state": "clean",
  "status": "ok",
  "type": "file"
}
{
  "action": "status",
  "bytesize": 115,
  "gitshasum": "519a1f93aa8b2965212e9287c6d8943a43952cdb",
  "parentds": "/cbica/projects/RBC/another_test",
  "path": "/cbica/projects/RBC/another_test/.gitattributes",
  "prev_gitshasum": "519a1f93aa8b2965212e9287c6d8943a43952cdb",
  "refds": "/cbica/projects/RBC/another_test",
  "state": "clean",
  "status": "ok",
  "type": "file"
}
{
  "action": "status",
  "bytesize": 5,
  "gitshasum": "1269488f7fb1f4b56a8c0e5eb48cecbfadfa9219",
  "parentds": "/cbica/projects/RBC/another_test",
  "path": "/cbica/projects/RBC/another_test/testfile",
  "prev_gitshasum": "1269488f7fb1f4b56a8c0e5eb48cecbfadfa9219",
  "refds": "/cbica/projects/RBC/another_test",
  "state": "clean",
  "status": "ok",
  "type": "file"
}
```
Does this look ok?

**Comment by kyleam** [comment]
Thanks.  Yeah, I don't see anything wrong there.  That makes me thing it is just the renderer and would be fixed by 025bc2ce0.

To be a little more confident:

```
$ python
Python 3.7.3 (default, Jul 25 2020, 13:03:44) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from os.path import relpath
>>> relpath(" " + "/cbica/projects/RBC/another_test/testfile", "/cbica/projects/RBC/another_test")
'../../../../home/kyle/ /cbica/projects/RBC/another_test/testfile'
>>> relpath("/cbica/projects/RBC/another_test/testfile", "/cbica/projects/RBC/another_test")
'testfile'
```

So, all good and should already be fixed in the maint branch.

**Comment by mattcieslak** [comment]
super helpful, thank you! Just to confirm on my system:

```
 [rbc@cubic-login2 ~]$ mkdir json_pp
 [rbc@cubic-login2 ~]$ cd json_pp
 [rbc@cubic-login2 json_pp]$ datalad -f json_pp create --description 'json renderer' .
[INFO   ] Creating a new annex repo at /cbica/projects/RBC/json_pp 
[INFO   ] Scanning for unlocked files (this may take some time) 
{
  "action": "create",
  "path": "/cbica/projects/RBC/json_pp",
  "refds": null,
  "status": "ok",
  "type": "dataset"
}
 [rbc@cubic-login2 json_pp]$ echo data >> testfile
 [rbc@cubic-login2 json_pp]$ datalad -f json_pp save -m 'save testfile' 
{
  "action": "add",
  "key": "MD5E-s5--6137cde4893c59f76f005a8123d8e8e6",
  "path": "/cbica/projects/RBC/json_pp/testfile",
  "refds": "/cbica/projects/RBC/json_pp",
  "status": "ok",
  "type": "file"
}
{
  "action": "save",
  "path": "/cbica/projects/RBC/json_pp",
  "refds": "/cbica/projects/RBC/json_pp",
  "status": "ok",
  "type": "dataset"
}
```

**Comment by kyleam** [comment]
I was still scratching my head about why some cases (like the script I gave above) don't show this issue.  In case anyone else was too, the answer is that `relpath` just happens to return the right result when `getcwd()` matches the reference dataset's path.

With a spurious leading space, `relpath` treats what was supposed to be an absolute path (`/a/b/`) as relative (`{cwd}/ /a/b`).  In the not uncommon case that the dataset matches `cwd`, the end result is ` /a/b`.  However, if `cwd` doesn't match the dataset, you get the `../` paths here.

<details>
<summary>reproducer with linked directory</summary>

```sh
set -eu
cd "$(mktemp -d "${TMPDIR:-/tmp}"/dl-XXXXXXX)"

mkdir a
ln -s a b
(
    cd b
    datalad create --description test .
    echo content > testfile
    datalad save -m 'initial commit'
)
```

Triggers issue on 0.13.5, resolved by maint's 025bc2ce0.

</details>
